pandas/core/tools/datetimes.py

```diff
@@ -316,7 +317,20 @@ def _convert_listlike_datetimes(
         if format is not None:
             raise ValueError("cannot specify both format and unit")
         arg = getattr(arg, "values", arg)
```

Comment:
this is the right idea, but needs to be done more generically, w/o peering inside EAs.

Reply:
change this
```arg = getattr(arg, "values", arg)```
to 
```arg = getattr(arg, "_values", arg)```

which will extract the EAs w/o conversion.

then create a new function where you do the conversion

Reply:
have a branch that checks if its an EA in an if/else
then use *only* the public functions, e.g. ``arr.isna()`` (which returns the mask)
use ``arr.na_value`` (for the NA value)
and use ``.take()`` to re-assemble.

Reply:
Hello. 

Sorry, I don't quite understand some of your suggestions. I made all the changes I knew how to do and tried to move from inner implementations to public functions. Specifically, I couldn't come up with a way to reassemble numpy arrays split by arbitrary indices using `.take()` , so I instead used `np.insert` to reassemble. This allowed me to stay away from specifying datatypes, preallocating an array, and assigning data by index.

Specific changes:
1.  Change `arg = getattr(arg, "values", arg)` to `arg = getattr(arg, "_values", arg)`. As far as I can tell, this makes no other changes necessary. Could you clarify?
2.  Switch from `arg._mask` to `arg.isna()`. Thanks.
3.  Switch to np.compress for grabbing the non-nan values from arg to generalize the code. Pass the result directly into `tslib.array_with_unit_to_datetime` to avoid storing it.
4.  Switch from preallocating an ndarray of datetime64 to using np.insert to inject NaT into the result.

Could you clear up what you meant by using `.take()` to re-assemble the result? I also avoided using `arr.na_value`, because I was only able to find it for pandas arrays.

Test results:
```
py.test pandas/tests/indexes/datetimes/test_tools.py -k 'test_intarray_to_datetime'
============================= test session starts =============================
platform win32 -- Python 3.7.3, pytest-5.3.1, py-1.8.0, pluggy-0.13.0
rootdir: D:\git_contrib\pandas\pandas, inifile: setup.cfg
plugins: hypothesis-4.53.2, cov-2.8.1, forked-1.1.2, xdist-1.30.0
collected 416 items / 415 deselected / 1 selected

pandas\tests\indexes\datetimes\test_tools.py .                           [100%]

====================== 1 passed, 415 deselected in 0.19s ======================
```